Integrate the wikitext task in the webapp

JulienVig opened the pull request:
- [x] Fixes #669. 
Fixed by preventing local caching

- [x] Fixes #673.  
Added an error message suggesting changing browser.
![Screenshot 2024-05-27 at 11 38 19](https://github.com/epfml/disco/assets/33122365/2aa22652-32dc-49df-b854-2775aef72b85)

- [x] Fixes #674 
Extended memory caching and saving for gpt-tfjs

![Screenshot 2024-06-11 at 17 23 36](https://github.com/epfml/disco/assets/33122365/718c03ea-ecd0-4d38-b541-4a5e96b3b128)

- [x] Rework the validator's test and inference method to enable users to stop and also fix tfjs memory leaks
![Screenshot 2024-06-13 at 16 35 53](https://github.com/epfml/disco/assets/33122365/7bdbf68a-f63e-4288-9935-968affbc24f3)

JulienVig committed:
Prevent tokenizer from trying to load model from filesystem

JulienVig committed:
Slightly refactor web memory

JulienVig committed:
Add UI error messages

JulienVig committed:
Add tensorBackend field to task training information

JulienVig committed:
Include tensorBackend type in the memory mechanism

JulienVig committed:
Remove stateful attributes from gpt-tfs to allow saving/loading models in the browser

JulienVig committed:
Throw error rather than toaster message

JulienVig committed:
Remove unused variable

JulienVig committed:
Add tensorbackend to tests

JulienVig committed:
Merge

JulienVig committed:
Remove git merge artifacts

JulienVig committed:
tensorBackend to skin condition task

tharvik commented on discojs-web/src/memory/memory.ts:
```diff
@@ -10,62 +10,72 @@ import { Map } from 'immutable'
     }
-
-    if (source.type === undefined || source.taskID === undefined || source.name === undefined) {
+    if (source.type === undefined || source.taskID === undefined ||
+      source.name === undefined || source.tensorBackend === undefined) {
```
`ModelSource` has only `type` & `version` as potentially undef, no need to check for everyone

```suggestion
    if (source.type === undefined) {
```

JulienVig replied:
I'll actually make `ModelInfo.type` required, I don't see any uses of it being left undefined right now

tharvik commented on discojs-web/src/memory/memory.ts:
```diff
@@ -10,62 +10,72 @@ import { Map } from 'immutable'
     const [name, versionSuffix] = fullName.split('@')
     const version = versionSuffix === undefined ? 0 : Number(versionSuffix)
-    return { type, taskID, name, version }
+    if (tensorBackend != 'tfjs' && tensorBackend != 'gpt') {
```
js is quite bad at comparing stuff, it loves to convert things if types differ. that's why `===` & `!==` where introduce, to not do that, as it creates unexcepted bugs in some cases. for example `{ toString: () => 'asd' } == 'asd'` is true which might come at a surprise.
anyway, I think that explicit is way more readable than implicit, or if the latter is used, a comment is welcome.

```suggestion
    if (tensorBackend !== 'tfjs' && tensorBackend !== 'gpt') {
```

JulienVig replied:
Ah yes indeed

tharvik commented on discojs-web/src/memory/memory.ts:
```diff
@@ -10,62 +10,72 @@ import { Map } from 'immutable'
+    const layersModel = await tf.loadLayersModel(this.getModelMemoryPath(source))
+    
+    const tensorBackend = this.getModelInfo(source).tensorBackend
+    if (tensorBackend == 'tfjs') {
+      return new models.TFJS(layersModel)
+    } else if (tensorBackend == 'gpt') {
+      return new models.GPT(undefined, layersModel)
+    } else {
+      throw Error("tensor backend unrecognized")
+    }
```
there is support for [exaustiveness checking](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#exhaustiveness-checking), which allow TS to know that every possible value was checked. a neat feature when changing the type of `tensorBackend` (as it will make compilation fail instead of throwing at runtime), and reads more nicely IMO

```suggestion
    switch (tensorBackend) {
      case 'tfjs':
        return new models.TFJS(layersModel)
      case 'gpt':
        return new models.GPT(undefined, layersModel)
    }
```

JulienVig replied:
Good to know thanks!

tharvik commented on discojs/src/memory/model_type.ts:
```diff
@@ -3,7 +3,7 @@
  * There can only be a single working model for a given task.
  */
-export enum ModelType {
+export enum StoredModelType {
```
while at it, we can simply have `export type StoredModelType = 'saved' | 'working'` instead of an enum (or even inlining it in `memory/base.ts` as it's in the only place where it's explicitly needed)

JulienVig replied:
After turning `StoredModelType` into a `type` I had to replace every call to `StoredModelType.{WORKING|SAVED}` by `'working'` and `'saved'`, was that the right way to do it?

tharvik replied:
yes, that's the typescript way™©®

tharvik commented on discojs/src/models/gpt/index.ts:
```diff
@@ -116,16 +117,22 @@ export class GPT extends Model {
 
-  [Symbol.dispose](): void{
-    console.log("Disposing model")
+  dispose() {
```
consider not having a public function for disposing, only the `[Symbol.dispose]()`. it pushes users to have a clear way to get rid of value, not two

JulienVig replied:
I find having a `dispose` method expected as it follows [tfjs API](https://github.com/tensorflow/tfjs/blob/fcad805d35a16efc4b9e42c9ef5950cf6c0693f2/tfjs-layers/src/engine/training.ts#L1916) but I guess we're trying to become backend agnostic... I'll remove it

tharvik commented on discojs/src/models/tokenizer.ts:
```diff
@@ -16,6 +16,7 @@ export async function getTaskTokenizer(task: Task): Promise<PreTrainedTokenizer>
   let tokenizer = task.trainingInformation.tokenizer
   if (tokenizer === undefined) throw Error('No tokenizer specified in the task training information')
   if (typeof tokenizer == 'string') {
+    env.allowLocalModels = false
```
hum, so that allow for loading of file-backed tokenizer, right? I don't see it used and might break in browsers (there is no filesystem, [kinda](https://developer.mozilla.org/en-US/docs/Web/API/File_System_API))

JulienVig replied:
It's actually needed in the browser, it disallows trying to load from the file system and enforces fetching on the huggingface remote hub.

tharvik replied:
ho, right, that's a default true for `allowLocalModels`, nvm then

JulienVig replied:
I'll add some doc actually it's quite an obscure line of code though very important

tharvik commented on discojs/src/task/training_information.ts:
```diff
@@ -57,6 +57,8 @@ export interface TrainingInformation {
   // truncate strings to a maximum length. The default value is tokenizer.model_max_length
   maxSequenceLength?: number
+  // Tensor framework used by the model
+  tensorBackend: 'tfjs' | 'gpt'
```
arf, I don't like adding stuff to `Task` and friends, but after looking more into it, it seems required until `Memory` is fixed to be able to store any type of model. too bad.

JulienVig replied:
Yes sadly I did this reluctantly...

tharvik commented on webapp/src/store/memory.ts:
```diff
@@ -33,11 +33,19 @@ export const useMemoryStore = defineStore('memory', () => {
         continue
       }
-      const [name, version] = fullName.split('@')
+      console.log(path, fullName)
```
debug?

tharvik commented on webapp/src/store/memory.ts:
```diff
@@ -33,11 +33,19 @@ export const useMemoryStore = defineStore('memory', () => {
+      console.log(path, fullName)
+      let name
+      let version
+      if (fullName == undefined) {
+        name = 'akjsd'
+        version = 0
```
that looks wrong, isn't `fullName` always defined?

JulienVig replied:
Ah good catch that's a remnant from transitioning some old models in memory to the new mechanism

tharvik commented on webapp/src/store/memory.ts:
```diff
@@ -58,10 +66,14 @@ export const useMemoryStore = defineStore('memory', () => {
         metadata.weightDataBytes,
       ].reduce((acc: number, v) => acc + (v === undefined ? 0 : v), 0)
 
+      if (tensorBackend != 'tfjs' && tensorBackend != 'gpt') {
```
```suggestion
      if (tensorBackend !== 'tfjs' && tensorBackend !== 'gpt') {
```

tharvik reviewed (approved):
wouhou, gpt in browser, well done!

testing the model is a bit weird, as there is not specific handler for that (got 5000% accuracy, such impressive model). but that's alright for now IMO, we can add it in a later iteration

JulienVig committed:
Clean code amd address PR comments

JulienVig commented:
Did you really get 5000% accuracy? If that's the case this is alarming, I don't see how it could happen

JulienVig committed:
Make StoredModelType type rather than enum

JulienVig committed:
Fix linting error

tharvik commented:
> Did you really get 5000% accuracy? If that's the case this is alarming, I don't see how it could happen

yep, it's fluctuating around 5000%±1000%, and use a huge amount of memory (before crashing my tab). the reproductability that I've is training wikitext once with "wiki.train.tokens" then testing it with "wiki.test.tokens". I'm using Firefox fwiw. before crashing it shows the next line in console:
`High memory usage in GPU: 1199.58 MB, most likely due to a memory leak`
and indeed, my system memory gets filled up quite rapidly (which triggers OOM, which kills the tab)

JulienVig commented:
Indeed I'll try to fix that before merging the PR, testing the model takes up more than 50GB

JulienVig committed:
Remove graph_informant

JulienVig committed:
Save transition between functional validator and while loop generator

JulienVig committed:
Make validator test generator (to allow stopping test) and fix memory leak

JulienVig committed:
Make validator inference a generator and fix memory leak

JulienVig commented:
@tharvik all fixed, I reworked the validator:
* removed the graph_informant (it was only used to query the validator accuracy)
* fixed the memory leak (tfjs sucks)
* fixed the logic mistake that yielded 5000 of accuracy
* renamed assess and predict into test and inference respectively
* made these methods generators to let users stop them. I had to turn the mapAsync into an iterator while loop to allow stopping, do you know if we could combine the functional way and generators somehow?

<img width=256 src="https://github.com/epfml/disco/assets/33122365/545ba55c-2a0c-4137-91a6-6b7aebbb0596">
<img width=256 src="https://github.com/epfml/disco/assets/33122365/e7844382-2026-43e8-8f2a-100260c7e374">

Until an LLM UI is implemented running inference is useless as it doesn't display anything

JulienVig committed:
Fix validator generator use

JulienVig committed:
Clean code

tharvik commented:
> @tharvik all fixed, I reworked the validator:

superbe, it tests nicely now :partying_face: (tfjs is shitty indeed)
and good idea to update the validator to be a real generator, that's more in line w/ training and more usuable.

> * made these methods generators to let users stop them. I had to turn the mapAsync into an iterator while loop to allow stopping, do you know if we could combine the functional way and generators somehow?

hum, no, you sadly have to have a while loop as `tf.data.Dataset` is not using generators. however, as datasets are lazy, you can have a dataset shaped with `mapAsync` and have a small while loop taking care of transforming it in a generator (maybe a bit more readable).
generators a quite powerful, but if libraries aren't exposing theses, it doesn't know how to stop processing a function. it can only `return` between lines, and in this case, there is a huge `await dataset.mapAsync.[...].toArray` so it can only stop before or after that.

> Until an LLM UI is implemented running inference is useless as it doesn't display anything

yep, I'm trying to draft up something basic in my PR, let's see what comes out of it :)
